homeassistant/components/geo_location/nsw_rural_fire_service_feed.py

```diff
@@ -0,0 +1,245 @@
+    categories = config.get(CONF_CATEGORIES)
+    # Initialize the entity manager.
+    NswRuralFireServiceFeedManager(hass, add_entities, scan_interval,
+                                   radius_in_km, categories)
+
+
+class NswRuralFireServiceFeedManager:
+    """Feed Manager for NSW Rural Fire Service GeoJSON feed."""
+
+    def __init__(self, hass, add_entities, scan_interval, radius_in_km,
+                 categories):
+        """Initialize the GeoJSON Feed Manager."""
+        from geojson_client.nsw_rural_fire_service_feed \
+            import NswRuralFireServiceFeed
+        self._hass = hass
+        self._feed = NswRuralFireServiceFeed((hass.config.latitude,
+                                              hass.config.longitude),
+                                             filter_radius=radius_in_km,
+                                             filter_categories=categories)
+        self._add_entities = add_entities
+        self._scan_interval = scan_interval
+        self._feed_entries = []
+        self._managed_entities = []
+        hass.bus.listen_once(
+            EVENT_HOMEASSISTANT_START, lambda _: self._update())
+        self._init_regular_updates()
+
+    def _init_regular_updates(self):
+        """Schedule regular updates at the specified interval."""
+        track_time_interval(self._hass, lambda now: self._update(),
+                            self._scan_interval)
+
+    def _update(self):
+        """Update the feed and then update connected entities."""
+        import geojson_client
+        status, feed_entries = self._feed.update()
+        if status == geojson_client.UPDATE_OK:
+            _LOGGER.debug("Data retrieved %s", feed_entries)
+            # Keep a copy of all feed entries for future lookups by entities.
+            self._feed_entries = feed_entries.copy()
+            keep_entries = self._update_or_remove_entities(feed_entries)
+            self._generate_new_entities(keep_entries)
+        elif status == geojson_client.UPDATE_OK_NO_DATA:
+            _LOGGER.debug("Update successful, but no data received from %s",
+                          self._feed)
+        else:
+            _LOGGER.warning("Update not successful, no data received from %s",
+                            self._feed)
+            # Remove all entities.
+            self._update_or_remove_entities([])
+
+    def _update_or_remove_entities(self, feed_entries):
+        """Update existing entries and remove obsolete entities."""
+        _LOGGER.debug("Entries for updating: %s", feed_entries)
+        remove_entry = None
+        # Remove obsolete entities for events that have disappeared
+        managed_entities = self._managed_entities.copy()
+        for entity in managed_entities:
+            # Remove entry from previous iteration - if applicable.
+            if remove_entry:
+                feed_entries.remove(remove_entry)
+                remove_entry = None
+            for entry in feed_entries:
```

Comment:
Nested for loops are not good. Let's convert the feed entries into a look up table:

```python
entries_lookup = {entry.external_id: entry for entry in feed_entries}
```

Reply:
Thanks. That had some good flow-on effects for simplifying the code.